shed/translation.py

```diff
@@ -0,0 +1,172 @@
+import uuid
+import time
+
+from streamz.core import Stream
+from databroker._core import ALL
+import networkx as nx
+
+
+def walk_up(node, graph, prior_node=None):
+    """Create graph from a single node, searching up and down the chain
+
+    Parameters
+    ----------
+    node: Stream instance
+    graph: networkx.DiGraph instance
+    """
+    if node is None:
+        return
+    t = hash(node)
+    graph.add_node(t, stream=node)
+    if prior_node:
+        tt = hash(prior_node)
+        if graph.has_edge(t, tt):
+            return
+        else:
+            graph.add_edge(t, tt)
+            if isinstance(node, FromEventModel):
+                return
+
+    for node2 in node.upstreams:
+        # Stop at translation node
+        if node2 is not None:
+            walk_up(node2, graph, node)
+
+
+class FromEventModel(Stream):
+    def __init__(self, upstream, doc_type, data_address, event_stream_name=ALL,
+                 stream_name=None, principle=False):
+        Stream.__init__(self, upstream, stream_name=stream_name)
+        self.stopped = False
+        self.principle = principle
+        self.doc_type = doc_type
+        if isinstance(data_address, str):
+            data_address = tuple([data_address])
+        self.data_address = data_address
+        self.event_stream_name = event_stream_name
+        self.start_uid = None
+        self.descriptor_uids = None
+        self.run_start_uid = None
+        self.subs = []
+
+    def update(self, x, who=None):
+        name, doc = x
+        if name == 'start':
+            self.stopped = False
+            self.start_uid = doc['uid']
+            self.descriptor_uids = {}
+        if name == 'descriptor':
+            self.descriptor_uids[doc['uid']] = doc.get('name', 'primary')
+        if name == 'stop':
+            [s._emit(s.create_stop(x)) for s in self.subs]
+        inner = doc.copy()
+        if (name == self.doc_type and
+                ((name == 'descriptor' and
+                  (self.event_stream_name is ALL or
+                   self.event_stream_name == doc.get('name', 'primary'))) or
+                 (name == 'event' and
+                  (self.event_stream_name == ALL or
+                   self.descriptor_uids[doc['descriptor']] ==
+                   self.event_stream_name)) or
+                 name in ['start', 'stop'])):
+            for da in self.data_address:
+                inner = inner[da]
+            return self._emit(inner)
+
+
+class ToEventModel(Stream):
```

Comment:
Again, some docstring would be useful. Here is what I gather:
Converts from a `Stream` back to an `EventStream`.

upstream : the upstream node
data_keys : the keys for the incoming data
    Note that the data is a assumed a tuple and len(data) = len(data_keys)